### DISCUSSION TITLE
Performance question

### DISCUSSION THREAD

**Discussion by gdborton** [vadimdemedes/ink#401] [General]
Hello!

Is there any documentation around how to achieve the best absolute performance out of ink?

I'm rendering an audio visualizer, which needs to update many times a second.
![image](https://user-images.githubusercontent.com/4172067/101266449-d8ffbb80-3703-11eb-965a-6f3e81425bc2.png)

Things I've already done to improve performance:
 - limit visualizer to 20 rows
 - return a single <Text/> component, managing each line as a string manually
 - color output at the line level instead of character level, memoize the styler based on max lines
 - Trim space character from end of lines (this was huge, should characters/lines that don't change still add to render time)

![image](https://user-images.githubusercontent.com/4172067/101266499-20864780-3704-11eb-9196-28f75860eb3f.png)

You can see the space characters that were rendered as they're in color as opposed to gray.  Could I use margin or padding to remove these? Would that help render performance?

I also tried memoizing each character of the visualizer, I was hoping that the high/low values wouldn't change much and this would prevent renders, but seems to have made things worse.

I haven't measured the attempted fps, but I'm guessing it's 60+.

Here is the Spectrum component:
https://github.com/gdborton/those-cheeks/blob/master/source/Spectrum.tsx#L57

**Comment by tom-sherman** [comment]
Note that Ink has a locked refresh rate of 30fps. I've opened a separate discussion about this here: https://github.com/vadimdemedes/ink/discussions/657
